providers/openfeature-provider-flagd/src/openfeature/contrib/provider/flagd/resolvers/process/flags.py

```diff
@@ -1,7 +1,42 @@
+import json
+import re
 import typing
 from dataclasses import dataclass
 
+from openfeature.event import ProviderEventDetails
 from openfeature.exception import ParseError
+from openfeature.provider.provider import AbstractProvider
+
+
+class FlagStore:
+    def __init__(
+        self,
+        provider: AbstractProvider,
+    ):
+        self.provider = provider
+        self.flags: typing.Mapping[str, "Flag"] = {}
+
+    def get_flag(self, key: str) -> typing.Optional["Flag"]:
+        return self.flags.get(key)
+
+    def update(self, flags_data: dict) -> None:
+        flags = flags_data.get("flags", {})
+        evaluators: typing.Optional[dict] = flags_data.get("$evaluators")
+        if evaluators:
+            transposed = json.dumps(flags)
+            for name, rule in evaluators.items():
+                transposed = re.sub(
+                    rf"{{\s*\"\$ref\":\s*\"{name}\"\s*}}", json.dumps(rule), transposed
+                )
+            flags = json.loads(transposed)
+
+        if not isinstance(flags, dict):
+            raise ParseError("`flags` key of configuration must be a dictionary")
+        self.flags = {key: Flag.from_dict(key, data) for key, data in flags.items()}
+
+        self.provider.emit_provider_configuration_changed(
+            ProviderEventDetails(flags_changed=list(self.flags.keys()))
+        )
```

Comment:
I would prefer to treat `AbstractProvider.emit_<event>` methods as protected. Can we have the provider listen to store changes and emit these events itself, to better separate responsibilities?

This would have the benefit of decoupling the FlagStore from a provider, making it easier to test.

Reply:
Toyed with this a bit, I can pass function refs instead of object refs around to decouple things better. But if we want to keep this logic in the provider itself we cannot depend on AbstractProvider and need to depend on FlagdProvider since it now has extra functions beyond the abstract protocol.

I'll commit an attempt but I think this is up to taste, let me know if you have a specific approach to suggest

Reply:
Check out this commit: https://github.com/open-feature/python-sdk-contrib/pull/84/commits/4b34fadf94b0796567d74d7b4bf00ca37d9be2ef

It's a bit of a compromise. Doesn't add extra listener interfaces (which I think harm the readability of the code) or any circular dependencies, but it does achieve the goal of removing the heavy dependency on `AbstractProvider` everywhere.